executor/write.go

```diff
@@ -38,23 +38,26 @@ var (
 	_ Executor = &LoadData{}
 )
 
-func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, assignFlag []bool, t table.Table, onDuplicateUpdate bool) (bool, error) {
-	cols := t.WritableCols()
-	touched := make(map[int]bool, len(cols))
-	assignExists := false
-	sc := ctx.GetSessionVars().StmtCtx
-	var newHandle types.Datum
-	for i, hasSetExpr := range assignFlag {
-		if !hasSetExpr {
-			if onDuplicateUpdate {
-				newData[i] = oldData[i]
-			}
-			continue
-		}
-		col := cols[i]
-		if col.IsPKHandleColumn(t.Meta()) {
-			newHandle = newData[i]
+// 1. fill values into on-update-now fields;
+// 2. rebase auto increment id if the field is changed;
+// 3. cast changed fields with respective columns;
+// 4. calculate it updates the record really or not;
+// 5. check not-null constraints.
+func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, touched []bool, t table.Table, onDup bool) (bool, error) {
+	var sc = ctx.GetSessionVars().StmtCtx
+	var changed = false
+	var handleChanged = false
+	var onUpdate = make([]bool, len(touched))
+
+	// We can iterate on public columns not writable columns,
+	// because all writable columns are after public columns.
```

Comment:
they are sorted by their `Offset`, and write-only columns' `Offset` must be larger than public columns.